app/src/main/java/org/jellyfin/androidtv/data/eventhandling/SocketHandler.kt

```diff
@@ -0,0 +1,164 @@
+import org.jellyfin.apiclient.model.entities.MediaType
+import org.jellyfin.sdk.api.client.ApiClient
+import org.jellyfin.sdk.api.client.extensions.sessionApi
+import org.jellyfin.sdk.api.client.extensions.userLibraryApi
+import org.jellyfin.sdk.api.sockets.SocketInstance
+import org.jellyfin.sdk.api.sockets.addGeneralCommandsListener
+import org.jellyfin.sdk.api.sockets.addListener
+import org.jellyfin.sdk.model.api.GeneralCommandType
+import org.jellyfin.sdk.model.api.LibraryUpdateInfo
+import org.jellyfin.sdk.model.api.PlaystateCommand
+import org.jellyfin.sdk.model.extensions.getValue
+import org.jellyfin.sdk.model.serializer.toUUIDOrNull
+import org.jellyfin.sdk.model.socket.LibraryChangedMessage
+import org.jellyfin.sdk.model.socket.PlayMessage
+import org.jellyfin.sdk.model.socket.PlayStateMessage
+import timber.log.Timber
+import java.util.UUID
+
+class SocketHandler(
+	private val context: Context,
+	private val api: ApiClient,
+	private val dataRefreshService: DataRefreshService,
+	private val mediaManager: MediaManager,
+) {
+	private val coroutineScope = CoroutineScope(Dispatchers.IO)
+	private var socketInstance: SocketInstance? = null
+
+	suspend fun updateSession() {
+		api.sessionApi.postCapabilities(
+			playableMediaTypes = listOf(MediaType.Video, MediaType.Audio),
+			supportsMediaControl = true,
+			supportedCommands = listOf(
+				GeneralCommandType.DISPLAY_MESSAGE,
+				GeneralCommandType.SEND_STRING,
+			),
+		)
+
+		if (socketInstance != null) socketInstance?.updateCredentials()
+		else socketInstance = createInstance()
+	}
+
+	private fun createInstance() = api.ws().apply {
+		// Library
+		addListener<LibraryChangedMessage> { message -> onLibraryChanged(message.info) }
+
+		// Media playback
+		addListener<PlayMessage> { message -> onPlayMessage(message) }
+		addListener<PlayStateMessage> { message -> onPlayStateMessage(message) }
+
+		// General commands
+		addGeneralCommandsListener(setOf(GeneralCommandType.DISPLAY_CONTENT)) { message ->
+			val itemId by message
+			val itemUuid = itemId?.toUUIDOrNull()
+
+			if (itemUuid != null) onDisplayContent(itemUuid)
+		}
+		addGeneralCommandsListener(setOf(GeneralCommandType.DISPLAY_MESSAGE, GeneralCommandType.SEND_STRING)) { message ->
+			val header by message
+			val text by message
+			val string by message
+
+			onDisplayMessage(header, text ?: string)
+		}
+	}
+
+	private fun onLibraryChanged(info: LibraryUpdateInfo) {
+		Timber.d(buildString {
+			appendLine("Library changed.")
+			appendLine("Added ${info.itemsAdded!!.size} items")
+			appendLine("Removed ${info.itemsRemoved!!.size} items")
+			appendLine("Updated ${info.itemsUpdated!!.size} items")
+		})
+
+		if (info.itemsAdded!!.any() || info.itemsRemoved!!.any())
+			dataRefreshService.lastLibraryChange = System.currentTimeMillis()
+	}
+
+	private fun onPlayMessage(message: PlayMessage) {
+		val itemId = message.request.itemIds?.firstOrNull() ?: return
+
+		PlaybackHelper.retrieveAndPlay(
```

Comment:
I do see this error in the logs but the audio still plays for me

Reply:
Ok this one is fixed now

Reply:
I tried again and it does play the audio now (the error is still there). It seems to have broken the remote controller though.

And if I then go click 'stop' in the remote control the app crashes due to it trying to call playbackController.endPlayback() and mFragment is null (line 1062, the closeActivity clause that was added in this PR). So it looks like it's not properly distinguishing between audio and video playback

Reply:
Do you have these issues with current master branch too? As I can't reproduce it with my 10.7.7 install.

Reply:
I've been comparing it to a build of the latest master, and everything work fine there.

with the latest commit to this PR, the error is gone and the commands work without crashes. However, now it just doesn't open the NowPlayingActivity. Also, it only ever queues 1 item. So if I try to play an album or artist it will play 1 song

Reply:
> However, now it just doesn't open the NowPlayingActivity

This is expected because the SocketHandler only has an ApplicationContext and we removed the currentactivity stuff.

> Also, it only ever queues 1 item. So if I try to play an album or artist it will play 1 song

When a queue needs to play we receive a list of item id's. But we don't have a function in the playbackhelper that takes those. That's why I opted to just play the first item.

Reply:
Comparing the old code in TVApiEventListener that handles the play command to the code in PlaybackHelper.retrieveAndPlay() -> PlaybackHelper.play() that's used by SocketHandler, it looks very similar, and both use the ApiClient. It looks like some of the TVApiEventListener code could be copied into PlaybackHelper to handle multiple items.

On this line it takes a list of IDs:
https://github.com/jellyfin/jellyfin-androidtv/blob/db3af6522654a269c21e3fe2023377b3d1700649/app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java#L179

Reply:
If we did copy the code it would use the apiclient, and I do not approve of adding new apiclient usages.

Reply:
Without a way to "translate" SDK items to be useable by the legacy playback code, and without compromising on using the apiclient to retain feature parity with the old implementation, would the consequence of merging this as-is be that remote control playing more than one item is broken until the playback rewrite is done?